### PR TITLE
Phosphorus requirements fix

### LINKED ISSUE #1588: Diet formulation for Phosphorus requirements results in unrealistically low Phosphorus feeding levels and manure phosphorus excretion
The current diet formulation methods are established to, at a minimum phosphorus requirements. However, this results in diets with lower phosphorus concentrations and lower total phosphorus intakes than are actually fed in practice. The downstream effects are that the amount of phosphorus excreted in the manure is also unrealistically low. 
## Expected Behavior

For lactating cows, we would expect dietary phosphorus concentrations to be 0.35% or higher which would result in a phosphorus intake of around 100g/cow/d or higher and a manure phosphorus excretion of ~60 g/cow/day or higher. 

## Current Behavior
Current pilot testing results from Julie after an update to the phosphorus requriement estimates were (look at the results from the Phos_req_fix): 
image

## Possible Solution

To achieve more realistic phophorus intake and subsequent manure excretion, we would like to introduce a requirement buffer variable that increases the value of the number used for the phosphorus requirement in the ration formulation methods by a certain percentage. This could be a new input in the animal input json or as a constant. Long term, I think it makes more sense as a user input but short term as a constant. 

## Steps to Reproduce

1. Run pilot farm c2f7 on the main branch using the following report filters: 

```

      {
          "name": "Pen 3 daily DMI per cow, kg/cow",
          "filters": ["AnimalModuleReporter.report_ration_interval_data.ration_nutrient_amount_pen_3_LAC_COW"],
          "variables": ["dm"],
          "vertical_aggregation": "average",
          "slice_start": -365
      }, 
      {
          "name": "Pen 3 daily phosphorus intake per cow, kg/cow",
          "filters": ["AnimalModuleReporter.report_ration_interval_data.ration_nutrient_amount_pen_3_LAC_COW"],
          "variables": ["phosphorus"],
          "vertical_aggregation": "average",
          "slice_start": -365
      },
      {
            "name": "Pen 3 dietary phosphorus, %",
            "cross_references":[
                "Pen 3 daily phosphorus intake per cow, kg/cow_phosphorus_ver_agg",
                "Pen 3 daily DMI per cow, kg/cow_dm_ver_agg"
            ],
            "constants": {
                "By 0.01": 0.01
            },
            "horizontal_aggregation": "division",
            "horizontal_order": [
                "Pen 3 daily phosphorus intake per cow, kg/cow_phosphorus_ver_agg",
                "Pen 3 daily DMI per cow, kg/cow_dm_ver_agg", 
                "By 0.01"
            ]
      }, 

      {
          "name": "Pen 3 daily manure phosphorus, g/cow/d",
          "filters": ["^AnimalModuleReporter.report_pen_manure_properties.pen_3_daily_phosphorus$",
                         "ManureHandlerDailyOutput_Pen_3_LAC_COW.num_animals"],
          "horizontal_aggregation": "division"
          "vertical_aggregation": "average",
          "horizontal_first": true,
          "horizontal_order": ["^AnimalModuleReporter.report_pen_manure_properties.pen_3_daily_phosphorus$",
                         "ManureHandlerDailyOutput_Pen_3_LAC_COW.num_animals"],
          "slice_start": -365
      },
```

## Context (Environment)

We are trying to get reasonable estimates for the phosphorus intake and manure phosphorus excretion to build the connection between animal diet and field phosphorus cycling

## Possible Implementation

In `animal_requirements.py` add a variable to the `calc_rqmts' method called `phosphorus_requirement_percent_buffer` or something similar to that. Instead of returning `phosphorus_requirement` return `phosphorus_requirement*phosphorus_requirement_buffer/100`

## Target Deadline

It would be great if this could be completed within 2 weeks

### PR DESCRIPTION
**Pull request by JoeWaddell**
### Context

Issue(s) closed by this pull request: closes #1588

- [x] The [changelog](https://docs.google.com/spreadsheets/d/1U-3igxdstHTFb6k5dVcE5-x9IL0r6q18eUiwwytKZyA/edit#gid=0) has been updated.

### What

Implemented a new value in the animal input JSON as suggested in Issue #1588

Corrected the `P_preg` section of phosophorus requirement calc to reflect the structure reported in the NASEM R code.

Corrected unit conversion per changes in #1583

Updated typing and docstrings, and unit test.

### Why

See tagged Issue: a new buffer value for Phosphorus intake was necessary to achieve realistic outputs. [Another followup PR ](https://github.com/RuminantFarmSystems/MASM/pull/1641) will make additional changes, if needed.

### Test plan

New buffer value methodology thoroughly examined by @KFosterReed , @jadamchick , and @krb78 using primarily (but not limited to) Pilot Test farm C2F7 using different buffer values.

Following report filter was used in analysis: 

      {
          "name": "Pen 3 daily DMI per cow, kg/cow",
          "filters": ["AnimalModuleReporter.report_ration_interval_data.ration_nutrient_amount_pen_3_LAC_COW"],
          "variables": ["dm"],
          "vertical_aggregation": "average",
          "slice_start": -365
      }, 
      {
          "name": "Pen 3 daily phosphorus intake per cow, kg/cow",
          "filters": ["AnimalModuleReporter.report_ration_interval_data.ration_nutrient_amount_pen_3_LAC_COW"],
          "variables": ["phosphorus"],
          "vertical_aggregation": "average",
          "slice_start": -365
      },
      {
            "name": "Pen 3 dietary phosphorus, %",
            "cross_references":[
                "Pen 3 daily phosphorus intake per cow, kg/cow_phosphorus_ver_agg",
                "Pen 3 daily DMI per cow, kg/cow_dm_ver_agg"
            ],
            "constants": {
                "By 0.01": 0.01
            },
            "horizontal_aggregation": "division",
            "horizontal_order": [
                "Pen 3 daily phosphorus intake per cow, kg/cow_phosphorus_ver_agg",
                "Pen 3 daily DMI per cow, kg/cow_dm_ver_agg", 
                "By 0.01"
            ]
      }, 
      {
          "name": "Pen 3 daily manure phosphorus, g/cow/d",
          "filters": ["^AnimalModuleReporter.report_pen_manure_properties.pen_3_daily_phosphorus$",
                         "ManureHandlerDailyOutput_Pen_3_LAC_COW.num_animals"],
          "horizontal_aggregation": "division"
          "vertical_aggregation": "average",
          "horizontal_first": true,
          "horizontal_order": ["^AnimalModuleReporter.report_pen_manure_properties.pen_3_daily_phosphorus$",
                         "ManureHandlerDailyOutput_Pen_3_LAC_COW.num_animals"],
          "slice_start": -365
      },

### CODE REVIEW

**Comment by KFosterReed** [comment]
The phosphorus buffer works! As long as the UDR functionality is turned off, increasing the buffer, increases the manure excretion and it seemed like, with this farm, I got reasonable values with a phosphorus buffer of 75%.

Here is a link to the simulations I tried, report filter used to generate the outputs attached here as well:
[report_phos_intake.json](https://github.com/RuminantFarmSystems/MASM/files/15182374/report_phos_intake.json)

[https://docs.google.com/spreadsheets/d/1CvuSNPz8yn1qp1F1GYP8ad90X3LpN8ut84Ui5d_fi0c/edit?usp=sharing](https://docs.google.com/spreadsheets/d/1CvuSNPz8yn1qp1F1GYP8ad90X3LpN8ut84Ui5d_fi0c/edit?usp=sharing)

Question for @jadamchick and @krb78: Given how the buffer on the requirements works in relation to the requirements, p-intake and p-excretion, do you think this a good choice for the method of adjustment? The only other alternative I can think of is that we set a fixed minimum requirement in g/d. That might be easier to use in some ways but theoretically it makes more sense to have the buffer implemented this way.

**Comment by JoeWaddell** [comment]
Great! Do you think it we should keep it (as it currently is in this PR) as an input, or do you think it's better to have it as a constant (as is commented out)?

For the second point: in the code we have a `MINIMUM_PHOSPHORUS` in `animal_module_constants.py` (set to 0.0 on main), but we could (instead of or in addition to this buffer) add an input so it's more modifiable on a farm by farm basis. Likewise we could set a minimum on the basis of animal type, but that starts to increase the number of inputs fairly rapidly.

**Comment by krb78** [comment]
Hi all! Sorry for the delay on this! I'm glad the phosphorus buffer works! It looks great and I agree with Kristan, it seems the 75% buffer gives the most realistic phosphorus excretions. It's a little odd to me that we need to feed so much higher percent of phosphorus in DM compared to what is truly being fed on dairy farms right now to achieve these outcomes, but so be it (for now). (I say that because it's my understanding that phosphorus is rarely above 0.40%, but I'm not a nutritionist so my view on this is limited). 

I think I prefer having the phosphorus implemented as a buffer that way we can dial it back if we'd like to, where if it's a constant, it's harder to manipulate from a user perspective.

**Comment by KFosterReed** [comment]
Great! Let's go with a single new user input of the phosphorus buffer percent! Thanks for the input Kaitlyn!

**Comment by jadamchick** [comment]
Thanks for this! I'm not sure why Github gives me a 404 error when I try to open Kristan's google sheets results, but definitely defer to you all on which scenario looks best. 
One question on using this: so should we change all herds to user_input = false (keep their ingredients by animal group, but let RuFaS balance the ration instead of using farm-defined inputs)?

**Comment by KFosterReed** [comment]
> Thanks for this! I'm not sure why Github gives me a 404 error when I try to open Kristan's google sheets results, but definitely defer to you all on which scenario looks best. One question on using this: so should we change all herds to user_input = false (keep their ingredients by animal group, but let RuFaS balance the ration instead of using farm-defined inputs)?

Sorry about the broken link Julie! Should be fixed now

**Comment by jadamchick** [comment]
> > Thanks for this! I'm not sure why Github gives me a 404 error when I try to open Kristan's google sheets results, but definitely defer to you all on which scenario looks best. One question on using this: so should we change all herds to user_input = false (keep their ingredients by animal group, but let RuFaS balance the ration instead of using farm-defined inputs)?
> 
> Sorry about the broken link Julie! Should be fixed now

Thanks Kristan! Probably my bad for waiting too long to open it :) works fine now

**Comment by jadamchick** [comment]
> Question for @jadamchick and @krb78: Given how the buffer on the requirements works in relation to the requirements, p-intake and p-excretion, do you think this a good choice for the method of adjustment? The only other alternative I can think of is that we set a fixed minimum requirement in g/d. That might be easier to use in some ways but theoretically it makes more sense to have the buffer implemented this way.

I agree that I think a buffer makes sense, and user-input adds nice access to be able to fiddle with it if needed; just would want there to be a good recommendation in the metadata for what the "strongly suggested starting value" should be or something similar.

**Comment by JoeWaddell** [comment]
Thanks all for the input and guidance on this: I'll remove the code and comments that pertain to the use of a constant. Likewise, I'll remove all reports and the input files on this branch before merging to main, just as a heads up in case any of that needs to be stored offline for later use!

I'll leave those files on this branch until approval however, since it'll require a SWE reviewer, and they might find it helpful.

**Comment by PooyaHekmati** [comment]
Why do we need to bring pilot farm data and metadata to main?
Also, what's up with the folder named other? It is out of place.

**Comment by JoeWaddell** [comment]
> Why do we need to bring pilot farm data and metadata to main?
> Also, what's up with the folder named other? It is out of place.

We don't! I added the `NOTE DURING REVIEW, CHANGES PRIOR TO MERGE` section in the PR description, as an explanation for those, and as a reminder to myself to remove those additional files prior to merge. Likewise I'll revert other changes to the default input files to reflect their current state on main. Just thought they could be useful during the review process.

**Comment by Isgotkowitz** [comment]
> Also, what's up with the folder named other? It is out of place.

I'm not sure about the "other" folder - but there is an "others" folder listed in the .gitignore, maybe the other here should actually be "others" and untracked.

File: RUFAS/routines/animal/life_cycle/animal_base.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/life_cycle/animal_base.py:114]: I think this should use `GRAMS_TO_KG` - because p_intake is already in grams.

File: RUFAS/routines/animal/ration/animal_requirements.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/animal_requirements.py]:
```suggestion
                1 + (AnimalBase.config["ration"]["phosphorus_requirement_buffer"] * GeneralConstants.PERCENTAGE_TO_FRACTION)
```

File: input/metadata/properties/default.json
Comment (Isgotkowitz) [inline input/metadata/properties/default.json:840]: Should the maximum be 100?

File: RUFAS/routines/animal/ration/ration_optimizer.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/ration_optimizer.py]: What is x?

File: RUFAS/routines/animal/ration/animal_requirements.py
Comment (Isgotkowitz) [inline RUFAS/routines/animal/ration/animal_requirements.py:1652]: I think it would be good to declare this 190 as a constant, either in this file or in the animal constants file.

File: tests/animal_module_tests/test_animal.py
Comment (Isgotkowitz) [inline tests/animal_module_tests/test_animal.py]:
```suggestion
    AnimalBase.config["ration"] = {"phosphorus_requirement_buffer": 35}
```

File: tests/animal_module_tests/test_animal.py
Comment (Isgotkowitz) [inline tests/animal_module_tests/test_animal.py]:
```suggestion
        "P_requirement": 1.35,
```
It is beneficial to hardcode expected results as much as possible in tests, because it will minimize the chances that an error in the logic being tested is carried over to the logic testing it.

**Review by Isgotkowitz — APPROVED** [review]
Left suggestions on some things that will help with clarity and test robustness, and had a question on one of the unit conversions but otherwise looks good.

**Comment by PooyaHekmati** [comment]
> > Also, what's up with the folder named other? It is out of place.
> 
> I'm not sure about the "other" folder - but there is an "others" folder listed in the .gitignore, maybe the other here should actually be "others" and untracked.

image

**Comment by PooyaHekmati** [comment]
> > Why do we need to bring pilot farm data and metadata to main?
> > Also, what's up with the folder named other? It is out of place.
> 
> We don't! I added the `NOTE DURING REVIEW, CHANGES PRIOR TO MERGE` section in the PR description, as an explanation for those, and as a reminder to myself to remove those additional files prior to merge. Likewise I'll revert other changes to the default input files to reflect their current state on main. Just thought they could be useful during the review process.

Making a PR means that the code is ready to be merged as-is. A PR has to have any temporary and test files and pieces of code removed, and a clean and ready-to-merge submission is needed. Adding things that facilitate the review process are definitely helpful and the right place for them is in the PR description. The Test Plan section captures how have you tested the code and know it is working. From what I understand, you have done more than just updating the test suite. Maybe you can rewrite the test plan section by something like run the code on xyz pilot farms data and metadata with XYZ filter and used ABC filter for output etc...

File: RUFAS/routines/animal/ration/ration_optimizer.py
Comment (JoeWaddell) [reply RUFAS/routines/animal/ration/ration_optimizer.py]: Looks like the original author forgot to finish this! It's the supply, I'll update it.

File: input/metadata/properties/default.json
Comment (JoeWaddell) [reply input/metadata/properties/default.json:840]: It is hypothetically unlimited, since it's an additional , if someone wanted to feed 3X additional amount, for instance, they would put 300. I've added a little to the description to explain it better. If it's still unclear I'm more than happy to rewrite it!

File: RUFAS/routines/animal/life_cycle/animal_base.py
Comment (JoeWaddell) [reply RUFAS/routines/animal/life_cycle/animal_base.py:114]: Corrected docstring: the attribute is in grams, but the input is in KG, so the code is now correct.

**Comment by jadamchick** [comment]
Here are more results using this branch on Zoetis pilot herds with the regional diet listed in table, user input = false. The 3rd column (lac cow P) is excreted phosphorus. 


<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/jms349/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/jms349/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72">


Herd | Diet | dietary phosphorus, %_hor_agg | LAC COW daily phosphorus intake   per cow, kg/cow_phosphorus_ver_agg | Lac   cow P, g/cow/d_hor_agg
-- | -- | -- | -- | --
21710_Q0 | FARMES23_West_default.json | 0.520329 | 0.149786 | 81.60258
21710_Q3 | FARMES23_West_default.json | 0.53295 | 0.155895 | 84.52568
21473_Q0 | FARMES23_West_default.json | 0.507271 | 0.137961 | 73.81164
21473_Q3 | FARMES23_West_default.json | 0.518483 | 0.144488 | 77.36521
19464_Q0 | FARMES23_Great Lakes_default.json | 0.505249 | 0.139255 | 74.28424
19464_Q3 | FARMES23_Great   Lakes_default.json | 0.530429 | 0.151369 | 79.92438
19001_Q0 | FARMES23_Upper Midwest_default.json | 0.533552 | 0.15782 | 85.35325
19001_Q3 | FARMES23_Upper   Midwest_default.json | 0.535902 | 0.160663 | 87.26133
18515_Q0 | FARMES23_Northeast_default.json | 0.490599 | 0.133296 | 73.4457
18515_Q3 | FARMES23_Northeast_default.json | 0.513123 | 0.145882 | 79.31144
18464_Q0 | FARMES23_Upper Midwest_default.json | 0.512792 | 0.144831 | 79.08955
18464_Q3 | FARMES23_Upper   Midwest_default.json | 0.516075 | 0.147255 | 80.60745
18257_Q0 | FARMES23_Upper   Midwest_default.json | 0.506162 | 0.140421 | 74.21006
18257_Q3 | FARMES23_Upper   Midwest_default.json | 0.521437 | 0.148798 | 79.16338
18114_Q0 | FARMES23_Upper   Midwest_default.json | 0.52815 | 0.155858 | 83.75007
18114_Q3 | FARMES23_Upper   Midwest_default.json | 0.538905 | 0.162506 | 87.54768
17990_Q0 | FARMES23_Upper   Midwest_default.json | 0.524519 | 0.146796 | 77.83421
17990_Q3 | FARMES23_Upper   Midwest_default.json | 0.54078 | 0.155634 | 82.60644
17928_Q0 | FARMES23_Intermountain_default.json | 0.502419 | 0.136574 | 73.61845
17928_Q3 | FARMES23_Intermountain_default.json | 0.50511 | 0.138602 | 74.10732
17721_Q0 | FARMES23_West_default.json | 0.513214 | 0.144773 | 77.42322
17721_Q3 | FARMES23_West_default.json | 0.516708 | 0.148436 | 79.83494
  |   |   |   |  



</body>

</html>

File: input/metadata/properties/default.json
Comment (Isgotkowitz) [reply input/metadata/properties/default.json:840]: No that makes perfect sense!

**Review by jadamchick — APPROVED** [review]
Looks good, I appreciate the description + example in the metadata - helpful context. Thanks Joe!

**Pull request merged**
